enhancements/kube-apiserver/bound-sa-tokens.md

```diff
@@ -0,0 +1,153 @@
+TokenRequest API.
+
+## Motivation
+
+3rd party IAM components such as the [AWS pod identity
+webhook](https://github.com/aws/amazon-eks-pod-identity-webhook) require bound tokens to
+be able to identify pods without the use of workarounds such as proxies. It is also
+generally desirable to support bound tokens to limit the scope of permissions (and
+therefore risk of compromise) for a given service account token.
+
+### Goals
+
+1. A bound token can be acquired via the TokenRequest API.
+2. A pod can request a bound service account token via a projected volume.
+3. The public key used to validate bound tokens can be retrieved for use with 3rd party IAM
+  components.
+
+### Non-Goals
+
+1. Integrate or ship the aws pod identity webhook
+2. Convert existing token-consuming components to use bound tokens
+
+## Proposal
+
+- Support configuring the options (via `KubeAPIServerConfig`) that will enable bound
+  tokens.
+  - The `TokenRequest` and `TokenRequestProjection` feature gates are enabled by default
+    in kube > 1.12, but are only configured by the apiserver if the following options are
+    provided:
+    - service-account-signing-key-file
+      - Operator should set this to the name of a private key it manages
+    - service-account-issuer
+      - Operator should set this to https://kubernetes.default.svc/
+    - api-audiences
+      - Operator should set this to https://kubernetes.default.svc/
+      - Bound tokens submitted to the apiserver must specify one or more of the audiences
+        configured here. If a request for a bound token does not specify an audience, the
+        audience will be defaulted to the issuer.
+  - Enabling bound tokens will satisfy goals #1 and #2.
+
+- The apiserver operator should manage a keypair used to sign and verify bound tokens
+  - The keypair needs to be distinct from the keypair used for legacy service account
+    tokens. Reusing the legacy token keypair would prevent token validation since tokens
+    are first validated by key and then by content. If the key were to validate, the
+    subsequent content check would fail due to differences in the content of legacy and
+    bound tokens.
+  - The keypair should be written to a secret in the `openshift-kube-apiserver` namespace
+  - The public key should be added to (rather than replaced in) a configmap in the
+    `openshift-config-managed` namespace.
+    - This configmap can be used to source the public keys needed by 3rd party components
+      to verify bound tokens issued by the apiserver, satisfying goal #3.
+    - The public keys in the configmap should be included in
+      `KubeAPIServerConfig.ServiceAccountPublicKeyFiles` for provision to the apiserver.
+      - In the event that the keypair secret is deleted, a new keypair will be generated
+        but tokens signed by the previous private key will continue to validate since the
+        corresponding public key will still be used to validate bound tokens.
```

Comment:
so the `openshift-config-managed` configmap has multiple public key possibly? How are they named?

Reply:
Updated to reflect: sequentially, in the same manner as sa token public keys are maintained.